### PR TITLE
make accommodations necessary for iOS 10 HLS

### PR DESCRIPTION
**Pull request by machenmusik**
Provides core support needed for iOS 10 HLS playback discussed in #1416.

Note that this leaves video flipped in Y dimension and also with B and R color channels reversed.
It is expected that corrections can be applied using custom shaders.
(Accommodating those corrections using the built-in shaders is outside the scope of this PR.)

### CODE REVIEW

**Comment by machenmusik** [comment]
(Due to the current Mobile Safari CORS issue, this is a bit of a PITA to diagnose / address / test as the A-Frame page and the HLS content must have the same domain.)

**Comment by dmarcos** [comment]
So with PR the video at least plays but flipped and with wrong color? :)

**Comment by dmarcos** [comment]
Is that better than the video not playing at all?

**Comment by machenmusik** [comment]
> Is that better than the video not playing at all?
Without this change, there is no way to make it work at all.
With this change, it is actually simple to use custom shader to fix flip and channel order.

So yes, this change is much better than the video not playing at all... see discussion in #1416

**Comment by dmarcos** [comment]
Can we add add a custom shader that it's set automatically when playing on iOS?

**Comment by machenmusik** [comment]
shader added but I haven't got it being set automatically yet

**Comment by machenmusik** [comment]
ok, the switching could probably use a little better guard (e.g. auto-switch if it's a shader like standard or flat we know won't work) if you wanted to touch up the PR

File: src/systems/material.js [hunk 1]
```diff
@@ -144,6 +144,16 @@ module.exports.System = registerSystem('material', {
     texture.minFilter = THREE.LinearFilter;
     setTextureProperties(texture, data);
 
+    // if we're on iOS, and the video is HLS, we currently need to do some hacks
+    if (utils.device.isIOS() && isHLS(videoEl)) {
+      // really it's BGRA, so this needs correction in shader
+      texture.format = THREE.RGBAFormat;
+      texture.needsCorrectionBGRA = true;
```
Comment (dmarcos) [inline src/systems/material.js]: Can we add this to the `this` instead of the texture?

File: src/systems/material.js [hunk 2]
```diff
@@ -144,6 +144,16 @@ module.exports.System = registerSystem('material', {
...
+      // apparently this is needed for HLS, so this needs correction in shader
+      texture.flipY = false;
+      texture.needsCorrectionFlipY = true;
```
Comment (dmarcos) [inline src/systems/material.js]: Can we add this to the this instead of the texture?

Comment (machenmusik) [inline src/systems/material.js]: (see above)

File: src/systems/material.js (hunk 1 above)
Comment (machenmusik) [inline src/systems/material.js]: I don't think so - `this` is the system, not a given video texture, right?  Complicating things is the fact that iOS 10 MP4 playback does not need any of these workarounds, so needs to be done on a per-video-texture basis.

**Comment by machenmusik** [comment]
due to CORS issues with Mobile Safari, in order to test this, your HTML page must be in the same exact domain as the IDevice-compatible video content (m3u8 manifests and ts / aac files),  Subdomains will not work at present.

**Comment by cvan** [comment]
> due to CORS issues with Mobile Safari, in order to test this, your HTML page must be in the same exact domain as the IDevice-compatible video content (m3u8 manifests and ts / aac files), Subdomains will not work at present.

can you point me to a sample video you're using as a test case?

File: src/systems/material.js [hunk 3]
```diff
@@ -195,6 +205,12 @@ module.exports.System = registerSystem('material', {
   }
 });
 
+function isHLS (videoEl) {
+  if (videoEl.type === 'application/x-mpegurl') return true;
```
Comment (cvan) [inline src/systems/material.js]: do MIME types get lowercased automatically when calling the `type` getter? 

I notice in [Apple's docs](https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/StreamingMediaGuide/DeployingHTTPLiveStreaming/DeployingHTTPLiveStreaming.html), it appears as `application/x-mpegURL`.

- also, do you want to check `vnd.apple.mpegURL` as well?
- and what about `video/MP2T` (`.ts`)?

File: src/systems/material.js [hunk 4]
```diff
@@ -195,6 +205,12 @@ module.exports.System = registerSystem('material', {
...
+  if (videoEl.src && videoEl.src.indexOf('.m3u8') > 0) return true;
```
Comment (cvan) [inline src/systems/material.js]: I would check `videoEl.src.toLowerCase()`

File: src/shaders/ios10hls.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
+var registerShader = require('../core/shader').registerShader;
+
+/**
+ * Custom shader for iOS 10 HLS.
```
Comment (cvan) [inline src/shaders/ios10hls.js]: can you add a comment with a URL to a doc/spec for this API?

**Comment by cvan** [comment]
@machenmusik nice work on supporting this 👍 if you can point me to some test case URLs, I'll try this out on my end.

File: src/systems/material.js (hunk 3 above)
Comment (machenmusik) [inline src/systems/material.js]: i've only seen lowercase, but good question.

not sure on your other two points

File: src/shaders/ios10hls.js (hunk 5 above)
Comment (machenmusik) [inline src/shaders/ios10hls.js]: sorry which API?

**Comment by machenmusik** [comment]
i think i used hls urls from the prior threejs tickets on the topic

maybe should just convert the aframe 360 video from MP4 to iPhone-safe HLS?

File: src/systems/material.js (hunk 4 above)
Comment (dmarcos) [inline src/systems/material.js]: Can we check for `toLowerCase` and add braces to the return?
````javascript
if (videoEl.src && videoEl.src.toLowerCase().indexOf('.m3u8') > 0) { return true; }
````

File: src/shaders/ios10hls.js (hunk 5 above)
Comment (cvan) [inline src/shaders/ios10hls.js]: some Apple docs (or Google's Web Fundamentals docs) about what HLS is

**Comment by cvan** [comment]
> maybe should just convert the aframe 360 video from MP4 to iPhone-safe HLS?

how about including both as `<source>`s?

**Comment by machenmusik** [comment]
if including both as `<source>` works, that's fine although it complicates the example since you can't do with inline `src` in `a-videosphere` for example -- I was just trying to say that if we want HLS source, should probably make a version of the recently-iDevice-friendly MP4 example that is HLS format

**Comment by cvan** [comment]
awesome, thanks for making the updates 👍 btw, is there a simple test or two you can add?

**Comment by wjhtinger** [comment]
How about play the live 360 stream from webrtc?

**Comment by cvan** [comment]
See [my comment](https://github.com/aframevr/aframe/issues/2663#issuecomment-300538257) in issue #2663. If @machenmusik has other insights, would be glad to know and discuss. Thanks, everyone 👍

**Comment by dmarcos** [comment]
@machenmusik I want to merge this but I would like to see it working first. Do you have any example I can try?

**Comment by machenmusik** [comment]
so per discussion with @dmarcos, it seems that recent changes may have broken this somehow.., may need to do a little more investigation before merging this one :-/

**Comment by machenmusik** [comment]
found the problem :-)

**Comment by dmarcos** [comment]
Thanks!

File: src/utils/material.js [hunk 6]
```diff
@@ -127,3 +131,9 @@ function handleTextureEvents (el, texture) {
   });
 }
 module.exports.handleTextureEvents = handleTextureEvents;
+
+module.exports.isHLS = function (videoEl) {
+  if (videoEl.type && videoEl.type.toLowerCase() in ['application/x-mpegurl', 'application/vnd.apple.mpegurl']) { return true; }
```
Comment (cvan) [inline src/utils/material.js]: arrays don't work like this. you've got to use `includes` or `indexOf` (which has better compatibility across the board). test for this?

File: src/utils/material.js [hunk 7]
```diff
@@ -127,3 +131,9 @@ function handleTextureEvents (el, texture) {
...
+  if (videoEl.src && videoEl.src.toLowerCase().indexOf('.m3u8') > 0) { return true; }
```
Comment (cvan) [inline src/utils/material.js]: what about checking for `.ts` too?

File: src/systems/material.js [hunk 8]
```diff
@@ -144,6 +145,16 @@ module.exports.System = registerSystem('material', {
     texture.minFilter = THREE.LinearFilter;
     setTextureProperties(texture, data);
 
+    // if we're on iOS, and the video is HLS, we currently need to do some hacks
+    if (utils.device.isIOS() && isHLS(videoEl)) {
```
Comment (cvan) [inline src/systems/material.js]: have you tested this on [Edge](https://webvr.rocks/microsoft_edge) (which supports WebVR in release channel)? Edge [supports HLS](http://caniuse.com/#feat=http-live-streaming) too; it may or may not have the same bugs.

Comment (machenmusik) [inline src/systems/material.js]: it does not.

File: src/utils/material.js (hunk 7 above)
Comment (machenmusik) [inline src/utils/material.js]: .ts is not HLS, it is not clear that either .ts playback works or that it needs the workaround (MP4 does not)

if you'd like to add .ts, and it needs this workaround, perhaps you'd like to do in separate PR as this one is merged now?

File: src/utils/material.js (hunk 6 above)
Comment (machenmusik) [inline src/utils/material.js]: maybe it only works based on src.   i do owe some tests (as subsequent PR) so will fix there as needed

File: src/systems/material.js (hunk 8 above)
Comment (cvan) [inline src/systems/material.js]: it does not have the bugs, or it does not have proper HLS support?

File: src/utils/material.js (hunk 6 above)
Comment (cvan) [inline src/utils/material.js]: it doesn't work for me in Chrome nor iOS - you sure?

screenshot of Chrome

[here's a test case](http://jsbin.com/qakosuhude) I tested on iOS.

Comment (machenmusik) [inline src/utils/material.js]: i think the test page i was using was matching on src (masking broken type check) - will fix when i add isHLS tests as mentioned above

Comment (machenmusik) [inline src/utils/material.js]: see #2668

File: src/systems/material.js (hunk 8 above)
Comment (machenmusik) [inline src/systems/material.js]: it does not appear to have the bugs.

File: src/utils/material.js (hunk 7 above)
Comment (cvan) [inline src/utils/material.js]: I'll test it some time to see - I just need to do some investigation there

File: src/utils/material.js (hunk 6 above)
Comment (cvan) [inline src/utils/material.js]: awesome! thanks
